packages/opentelemetry-exporter-stackdriver-trace/src/trace.ts

```diff
@@ -0,0 +1,144 @@
+import { NoopLogger } from '@opentelemetry/core';
+import { ReadableSpan, SpanExporter } from '@opentelemetry/tracing';
+import { Logger } from '@opentelemetry/types';
+import { GoogleAuth } from 'google-auth-library';
+import { google } from 'googleapis';
+import { StackdriverExporterOptions } from './external-types';
+import { getReadableSpanTransformer } from './transform';
+import { Span, SpansWithCredentials } from './types';
+
+google.options({ headers: { 'x-opentelemetry-outgoing-request': 1 } });
+
+/**
+ * Format and sends span information to StackDriver Trace.
+ */
+export class StackdriverTraceExporter implements SpanExporter {
+  private _projectId: string | void | Promise<string | void>;
+  private readonly _logger: Logger;
+  private readonly _auth: GoogleAuth;
+
+  private static readonly _cloudTrace = google.cloudtrace('v2');
+
+  constructor(options: StackdriverExporterOptions) {
+    this._logger = options.logger || new NoopLogger();
+
+    this._auth = new GoogleAuth({
+      credentials: options.credentials,
+      keyFile: options.keyFile,
+      keyFilename: options.keyFilename,
+      projectId: options.projectId,
+      scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+
+    // Start this async process as early as possible. It will be
+    // awaited on the first export because constructors are synchronous
+    this._projectId = this._auth.getProjectId().catch(err => {
+      this._logger.error(err);
+    });
+  }
+
+  /**
+   * Publishes a list of spans to Stackdriver.
+   * @param spans The list of spans to transmit to Stackdriver
+   */
+  async export(
+    spans: ReadableSpan[],
+    resultCallback: (result: ExportResult) => void
+  ): Promise<void> {
+    if (this._projectId instanceof Promise) {
+      this._projectId = await this._projectId;
+    }
+
+    if (!this._projectId) {
+      return resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+    }
+
+    this._logger.debug('StackDriver Trace export');
+    const authorizedSpans = await this._authorize(
+      spans.map(getReadableSpanTransformer(this._projectId))
+    );
+
+    if (!authorizedSpans) {
+      return resultCallback(ExportResult.FAILED_NOT_RETRYABLE);
+    }
+    this._logger.debug('StackDriver Trace got span authorization');
+
+    try {
+      await this._batchWriteSpans(authorizedSpans);
+      resultCallback(ExportResult.SUCCESS);
+    } catch (err) {
+      this._logger.error(`Stackdriver Trace failed to export ${err}`);
+      resultCallback(ExportResult.FAILED_RETRYABLE);
+    }
+  }
+
+  shutdown(): void {}
+
+  /**
+   * Sends new spans to new or existing traces in the Stackdriver format to the
+   * service.
+   * @param spans
+   */
+  private _batchWriteSpans(spans: SpansWithCredentials) {
+    this._logger.debug('StackDriver Trace batch writing traces');
+
+    return new Promise((resolve, reject) => {
+      // TODO: Consider to use gRPC call (BatchWriteSpansRequest) for sending
```

Comment:
Nit: @TODO ?

Reply:
we must have different editor extensions :)

i'll update it since it seems like more people use `@todo`